implementation-details-overlay.md

```diff
@@ -0,0 +1,340 @@
+
+## E.1 - Sup Protocol Specific Bootnodes
+
+
+## E.1 - Bootstrapping via Bootnodes
+
+The client uses a set of bootnodes to acquire an initial view of the network.
+
+### E.1.a - Bootnodes
+
+Each supported sub protocol can have its own set of bootnodes.  These records can be either hard coded into the client or provided via client configuration.
+
+## E.2 - Population of routing table
+
+The client actively seeks to populate its routing table by performing [RFN](#TODO) lookups to discover new nodes for the routing table
+
+## E.3 - Liveliness checks
+
+The client tracks *liveliness* of nodes in its routing table and periodically checks the liveliness of the node in its routing table which was least recently checked.
+
+### E.3.a - Rate Limiting Liviliness Checks
+
+The liviliness checks for any individual node are rate limited as to not spam individual nodes with lots of PING messages when the routing table is sparse.
+
+# F - Content Database
+
+Management of stored content.
+
+## F.1 - Content can be stored
+
+Content can be stored in a persistent database.  Databases are segmented by sub protocol.
+
+## F.2 - Content can be retrieved by `content_id`
+
+Given a known `content_id` the corresponding content payload can be retrieved.
+
+## F.3 - Content can be removed
+
+Content can be removed.
+
+
+## F.3 - Query furthest by distance
+
+Retrieval of the content from the database which is furthest from a provided `node_id` using the custom distance function.
+
+
+## F.4 - Total size of stored content
+
+Retrieval of the total number of bytes stored.
+
+# G - Content Management
+
+## G.1 - Support for the uTP Sub Protocol
+
+Support for sending and receiving streams of data using the uTP sub protocol.
+
+### G.1.a - Support for outbound streams
+
+The ability to establish a new outbound connection with another node with a specified `connection-id`
+
+### G.1.a - Support for inbound streams
+
+The ability to listening for an inbound connection fromm another node with a `connection-id` that is known in advance.
+
+## G.2 - Enforcement of maximum stored content size
+
+When the total size of stored content exceeds the configured maximum content storage size the content which is furthest from the local `node_id` is evicted in a timely manner.
+
+## G.3 - Retrieval via FINDCONTENT/FOUNDCONTENT & uTP
+
+Support for retrieving content using the FINDCONTENT, FOUNDCONTENT, and uTP sub protocol.
+
+### G.3.a - DHT Traversal
+
+The client can use the FINDCONTENT and FOUNDCONTENT messages to traverse the DHT until they encounter a node that has the desired content.
+
+### G.3.b - Receipt via direct payload
+
+Upon encountering a FOUNDCONTENT response that contains the actual content payload, the client can return the payload.
+
+### G.3.b - Receipt via uTP
+
+Upon encountering a FOUNDCONTENT response that contains a uTP `connection-id`, the client can listen for and receive the content payload over the uTP stream.
```

Comment:
This needs to be better defined in the actual specs but perhaps it makes a little more sense that that the receiver of the `FOUNDCONTENT` message initiates the uTP connection? As the sender will not know when the message is received and thus when he can transmit the uTP SYN packet (Timing issues, might have to retry few times).

So this is a TODO for the portal-wire specs as currently direction is not specified there for this message type.

Reply:
Good catch.  Yes, initiation from the requester side seems to make the most sense here.